### DISCUSSION TITLE
Use several different custom interpolation functions simultaneously

### DISCUSSION THREAD

**Discussion by aaaaaaqing** [idaholab/moose#25046] [Q&A General]
In the finite element, the interpolation function (test function) is very important, especially in some algorithms, the selection of the interpolation function directly affects the results of the algorithm. At present, I want to use multiple interpolation functions in the same formula. So how do you use multiple interpolating functions in an equation at the same time (even scalar interpolating functions and vector interpolating functions at the same time)? If there is a way or a similar example, I would appreciate your help

**Comment by lindsayad** [comment]
Can you be more specific about what kind of discretization you want to perform? We do things like Streamline-Upwind Petrov-Galerkin in several places in the repository

**Reply by aaaaaaqing** [reply]
Thanks for your reply. I had some difficulties with moose's interpolation functions, first I wanted to use both vector and scalar functions to dediscretize equations, and to custom-build Jacobian matrices and residuals (both interpolation functions are used). But I don't know how to use this algorithm in moose, so I'd like to check if there are any good suggestions

**Reply by lindsayad** [reply]
That's not very clear to me. Do you have multiple physics? What are they?

**Reply by aaaaaaqing** [reply]
This is because the weak form I use comes from an algorithm:

I have a current continuity equation: ∇ jn-∂ n/∂t-Rn=0, where Jn is the current density, n is the carrier concentration, and Rn is the net recombination velocity. In the steady state case, this equation becomes ∇ Jn-Rn =0, then use Jn = W (j) *Jn (j), where W is the vector interpolation function, then i use Ni as the scalar test function, then the equation becomes: N (i) *∇ (W (j) *Jn (j)) -Ni *Rn =0. Therefore, I want to use both interpolation functions at the same time, and whether moose is capable of handling this situation

**Reply by aaaaaaqing** [reply]
This is just an example that I'm giving you to show you my discretization, you don't have to think about the rationality of this algorithm, just to show why I want two different kinds of interpolation functions to be used at the same time, right

**Reply by lindsayad** [reply]
I'll think about the rationality of it as I see fit for answering the questions you are asking

**Reply by roystgnr** [reply]
By "two different kinds of interpolation functions" do you mean different Finite Element spaces?  You can certainly have both a vector-valued and a scalar-valued variable coupled into the same system of equations (though vector-valued spaces are tricky enough that in general users tend to construct them from tensor products of scalar spaces unless they have some special needs like Nedelec elements where the space can't be decomposed that way).

Or are you trying to talk about Petrov-Galerkin methods?  Your question here is really hard to parse.

**Reply by cticenhour** [reply]
To add on to @roystgnr's comment, we do have examples of multiple-variable systems with multiple finite element types (one being scalar, and another being vector). One example is here: https://github.com/idaholab/moose/blob/next/test/tests/kernels/vector_fe/coupled_scalar_vector.i
